### ISSUE TITLE
Draft: Revise HTML for Tabbed extension

### ISSUE THREAD

**Issue by squidfunk** [facelessuser/pymdown-extensions#1415]
__TL;DR: this is a draft to improve on the current implementation of the Tabbed extension. It is by no means ready for implementation. It should serve as a base for further discussion to learn whether we can gravitate towards a better solution than we currently have. There are still problems with this approach__.

## The problem

The Tabbed extension is pretty awesome and many users love it. It provides a lot of value, especially to Material for MkDocs. However, it suffers from some problems that are not solvable without additional JavaScript. The main problem is that on narrower screen sizes, tabs are broken onto separate lines, like here for example "Hide both":

<img width="366" alt="Bildschirmfoto 2021-08-06 um 22 15 23" src="https://user-images.githubusercontent.com/932156/128566488-f37bbfd4-c5c1-4a1a-bbd8-68ae59afa6ae.png">

This cannot be mitigated, as the tabs markup is defined as follows:

``` html
<container>

  <!-- 1st tab -->
  <input 1 />
  <label 1 />
  <content 1 />

  <!-- 2nd tab -->
  <input 2 />
  <label 2 />
  <content 2 />

  ...
</container>
```

In order to make the labels overflow and scrollable, they would need to be located inside a container. With my knowledge of HTML and CSS, I'm very certain that there's no way to solve this problem without changing the underlying HTML. Furthermore, this would break the current tabs activation approach which relies entirely on co-location of `input` and `label` elements.

## A solution attempt

I've re-architected the tabs HTML and found a solution that allows us to overflow the tabs label container. It looks like this:

<img width="459" alt="Bildschirmfoto 2021-08-06 um 22 22 05" src="https://user-images.githubusercontent.com/932156/128567099-63b3e0fe-3375-4604-ac8f-eaee9a20ba60.png">

https://user-images.githubusercontent.com/932156/128594641-00fb9e53-facc-43f2-8a2f-79c418baad27.mp4

The HTML can be found here, so just drop it into a `*.html` file and play with it:

<details>
<summary><b>Show revised HTML</b></summary>

``` html
<html>
  <head>
    <style>

      /* --- Boilerplate --- */

      body {
        font-family: sans-serif;
        margin: 40px;
      }

      body > label {
        display: inline-block;
        padding: 0px 0px 20px;
        cursor: pointer;
      }

      article {
        box-shadow: 0 0 10px black;
        overflow: hidden;
      }

      :checked ~ article {
        width: 400px;
      }

      /* --- Tab label --- */

      /* Tab input */
      .tabbed-set input {
        position: absolute;
        width: 0;
        height: 0;
        opacity: 0;
      }

      /* Tab label container */
      .tabbed-labels {
        display: flex;
        overflow: auto;
        box-shadow: 0 -1px 0 #CCC inset;
        scroll-snap-type: x proximity;
      }

      /* Tab label */
      .tabbed-labels > * {
        display: inline-block;
        padding: 10px;
        white-space: nowrap;
        scroll-snap-align: start;
        cursor: pointer;
      }

      /* Hide scrollbar for Chrome, Safari and Opera */
      .tabbed-labels::-webkit-scrollbar {
        display: none;
      }

      /* Hide scrollbar for IE, Edge and Firefox */
      .tabbed-labels {
        -ms-overflow-style: none;  /* IE and Edge */
        scrollbar-width: none;  /* Firefox */
      }

      /* Tab label is active */
      .tabbed-set input:nth-child(1):checked ~ .tabbed-labels > :nth-child(1),
      .tabbed-set input:nth-child(2):checked ~ .tabbed-labels > :nth-child(2),
      .tabbed-set input:nth-child(3):checked ~ .tabbed-labels > :nth-child(3) {
        color: deeppink;
        box-shadow: 0 -4px 0 deeppink inset;
      }

      /* --- Tab content --- */

      /* Inactive tab */
      .tabbed-content div {
        display: none;
      }

      /* Active tab */
      .tabbed-set input:nth-child(1):checked ~ .tabbed-content > :nth-child(1),
      .tabbed-set input:nth-child(2):checked ~ .tabbed-content > :nth-child(2),
      .tabbed-set input:nth-child(3):checked ~ .tabbed-content > :nth-child(3) {
        display: block;
        padding: 10px;
      }
    </style>
  </head>
  <body>

    <!-- Just for the demo -->
    <input type="checkbox" id="limit" checked>
    <label for="limit">Limit width of container</label>

    <!-- The interesting stuff -->
    <article>
      <div class="tabbed-set" data-tabs="10:3">
        <input id="__tabbed_10_1" name="__tabbed_10" type="radio" checked>
        <input id="__tabbed_10_2" name="__tabbed_10" type="radio">
        <input id="__tabbed_10_3" name="__tabbed_10" type="radio">
        <div class="tabbed-labels">
          <label for="__tabbed_10_1">A very long title</label>
          <label for="__tabbed_10_2">Another very long title</label>
          <label for="__tabbed_10_3">Well, why can't you just use shorter titles?</label>
        </div>
        <div class="tabbed-content">
          <div><p>Some stuff</p></div> 
          <div><p>More stuff</p></div> 
          <div><p>Even more stuff</p></div> 
        </div>
      </div>
    </article>
    <script>
      const tabs = document.querySelectorAll(".tabbed-set > input")
      for (const tab of tabs) {
        tab.addEventListener("change", () => {
          const label = document.querySelector(`label[for=${tab.id}]`)
          label.scrollIntoView({ behavior: "smooth" })
        })
      }
    </script>
  </body>
</html>
```

</details>

Here's how it works:

1. Labels and contents are grouped inside separate containers, allowing for better control of what is shown
2. `input` + `label` elements are targetted with `:nth-child(...)` selectors. While this seems unnecessarily bloated it is, AFAIK, the only viable approach. Furthermore, when transferred over the wire, it should compress very well, because the markup is largely the same, so `gzip` and friends should work very efficiently. We can provide, let's say, 10 selectors. That should be sufficient for 99.9% of all use cases. The CSS can be extended if more is necessary.

## Constraints

1. The print view of Material for MkDocs currently just expands all tabs and renders them below each other. A possible solution for this could be to add the tabs' title as an attribute to the content element and use a pseudo-element to render a label before the content container. I haven't tested this thoroughly, but it might be a start:

    HTML:

    ``` html
    <div class="tabbed-content">
      <div data-title="A very long title"><p>Some stuff</p></div> 
      ...
    </div>
    ```

    CSS:

    ``` css
    @media print {
      .tabbed-content > ::before {
        content: attr(data-title);
        ...
      }
    }
    ```

    Downside: some duplicate HTML (i.e. the tabs title), but personally, I could live with that.

2. We definitely shouldn't start relying on JavaScript for this functionality. The current CSS-only solution is pretty awesome, because it works on slow connections and when the site is partly ready due to missing JavaScript. For a comparison, try [Docusaurus's tabs implementation](https://docusaurus.io/docs/installation#running-the-development-server) after disabling JavaScript - it doesn't work at all.

3. We might need to keep the old implementation for older clients for some time.

---

Any feedback is greatly appreciated. If you don't wish to go down this path for whatever reason, feel free to close this issue. I've had this on my mind for a long time and wanted to give it a shot and publish it to get some feedback and ideas. I don't expect this to land quickly, since there is some stuff to work out.

Also, if GitHub issues is not the right place to discuss, feel free to move this to a discussion.

**Comment by facelessuser** [comment]
> Also, if GitHub issues is not the right place to discuss, feel free to move this to a discussion.

I think this is fine as it is a feature request "improve tabs". How to get there is a perfect discussion for this issue.

> The main problem is that on narrower screen sizes, tabs are broken onto separate lines, like here for example "Hide both":

Personally, I've never had a problem with the old behavior, but I'm open to exploring alternatives to see how it feels.

On the one hand, the old way is nice because you can see all the tabs no matter what. I imagine with the new way, you can have tabs overflow in such a way that you don't know there are overflows, also you have to scroll around to get to them, but I can see how it can be perceived to be nicer as it keeps the header contained to a single line.

I haven't looked at the HTML yet, but I'll probably get a chance this weekend to explore it more. I'm interested to try it out and see how it feels.

**Comment by facelessuser** [comment]
@gir-bot remove S: triage
@gir-bot add T: feature, S: needs-decision

**Comment by squidfunk** [comment]
> On the one hand, the old way is nice because you can see all the tabs no matter what. I imagine with the new way, you can have tabs overflow in such a way that you don't know there are overflows, also you have to scroll around to get to them, but I can see how it can be perceived to be nicer as it keeps the header contained to a single line.

I think my initial proposal which we used as a basis for implementation rendered them as an accordion on mobile, which might also be something to consider, but IMHO it adds too much vertical space, especially when you have many instances on one page. We could indicate that the container is scrollable to signal that to the user.

I'm not completely sold, so let's let that sink in for a while and maybe implement or abandon it. Nonetheless, it might be a viable option that is now documented here.

**Comment by facelessuser** [comment]
> I'm not completely sold, so let's let that sink in for a while and maybe implement or abandon it. Nonetheless, it might be a viable option that is now documented here.

Yep, understood. I am completely fine with exploring. It's one of those things that I think would be best experienced before really weighing in on whether it is good or not. I think a prototype is going to be appropriate. I was mainly just giving my initial thoughts, not really a definitive opinion.

**Comment by squidfunk** [comment]
I've update the HTML and video example:

1. Tab labels will now automatically come into view when clicked or focused (via keyboard) with a small animation
2. Tab labels will now snap to grid

Some further ideas:

- Leave a little space to the left and right when the tab label container overflows, so the next label is visible
- Add chevrons left and right, when the tab label container can be scrolled

This should address at least some of the concerns you raised.

**Comment by facelessuser** [comment]
I see in the prototype that you've added a shadow border and provided padding around the content. Is the intent moving forward to make them feel more like a container as opposed to inline alternate content? Or is that just to illustrate the boundaries in this test example? The padding, not the shadow border, is something that I've always done in the past as a personal modification. This is more a question of curiosity, not really a criticism or suggestion.

Looking at the HTML structure, it will probably be a bit more complex to handle for sure. I need to take a look at the source and see how complex it would be to mirror the structure you are proposing.

**Comment by squidfunk** [comment]
> I see in the prototype that you've added a shadow border and provided padding around the content. Is the intent moving forward to make them feel more like a container as opposed to inline alternate content? Or is that just to illustrate the boundaries in this test example?

Correct, that was just to better illustrate the enclosing container for the demo. In the end, they should look the same as they currently do.

> Looking at the HTML structure, it will probably be a bit more complex to handle for sure. I need to take a look at the source and see how complex it would be to mirror the structure you are proposing.

The HTML and CSS classes are not final. If you think it's generally feasible, we can discuss semantics.

**Comment by facelessuser** [comment]
So, I was playing around with some rough code (not confident that all corner cases are handled yet), but here is my concern. This is not generalized. You have to specify every `nth-child` combo. Is there a way to handle any number of inputs?

```css
/* Active tab */
      .tabbed-set input:nth-child(1):checked ~ .tabbed-content > :nth-child(1),
      .tabbed-set input:nth-child(2):checked ~ .tabbed-content > :nth-child(2),
      .tabbed-set input:nth-child(3):checked ~ .tabbed-content > :nth-child(3) {
        display: block;
        padding: 10px;
      }
```

**Comment by squidfunk** [comment]
No, that is one downside of this approach, but I think it's bearable. As I said, it should compress pretty well with `gzip`.

**Comment by facelessuser** [comment]
Hmm, so you have to pick some arbitrarily larger number, and if people exceeded that, too bad.

**Comment by squidfunk** [comment]
I think 10 should be enough, as I wrote in my original post.

**Comment by facelessuser** [comment]
> I think 10 should be enough, as I wrote in my original post.

Ah, yeah, I forgot about that part. Yeah, not quite as flexible as the original, but 10 feels like a lot of tabs 🙂.

**Comment by facelessuser** [comment]
I'm running through existing test cases. As long as there are no unexpected crazy difficult edge cases, I imagine I'll throw up an experimental branch that you can play with locally.

**Comment by facelessuser** [comment]
Just an FYI, I'll be wrapping the individual contents in `tabbed-subcontent` classes. It just makes my life a little easier when I'm constructing the tab sets. There's a lot of times you have to be aware of whether you are already in a tabset content div or if you have to find it to continue adding to the etc. Maybe that helps with CSS, maybe it doesn't, but it's going to be there regardless 🙃.

```html
<div class="tabbed-content">
            <div class="tabbed-subcontent">
            <p>whatever</p>
            </div>
           
            <div class="tabbed-subcontent">
            <p>more</p>
            </div>
</div>
```

**Comment by facelessuser** [comment]
I wouldn't count on it until maybe Sunday at the earliest🙂. I need to abstract old logic and new logic as right now I am just wholesale replacing the old logic. I should be done verifying tests later today. We'll just see how much I get through.

**Comment by squidfunk** [comment]
Sounds great! Im happy to try the experimental branch and provide an experimental implementation as part of Material for MkDocs. Maybe we could set another discernible CSS class on the outermost container, so I can use that for targeting to add experimental support on Material’s default branch and we can push out a POC for users to try.

**Comment by facelessuser** [comment]
Yeah, I may tag it with `tabbed-alternate` or something.

**Comment by facelessuser** [comment]
Got it done much sooner than I thought: https://github.com/facelessuser/pymdown-extensions/tree/feature/tabbed-alternate. Use the option `alternate_style` (I may call it something else in the final version -- assuming we move forward with this).

**Comment by squidfunk** [comment]
Perfect, thanks for implementing this so quickly! I'll add support for Material for MkDocs asap. If we decide that it looks and feels good, we could move forward in the following way:

1. A new version of Pymdown Extensions could add experimental support behind the `alternate_style` option
2. Material for MkDocs will add support for this option with a new version and bump its minimum `pymdownx` version req
3. We advertise it in the docs and see whether users run into any problems and test it in the wild
4. If all goes well, we might deprecate/remove the old style (assuming you don't want to support two styles). If not, we might just remove the experimental support again and be back to square 1.

**Comment by squidfunk** [comment]
I've finished the implementation, it was quite straight forward: https://github.com/squidfunk/mkdocs-material/pull/2915

**Comment by facelessuser** [comment]
I'm open to new CSS class names. If you have suggestions, feel free to make suggestions.

As far as `data-title`, I can look into it. I was focused on getting basic functionality that I completely forgot about it.

**Comment by facelessuser** [comment]
Before I attempt to add `data-title`, some titles can have things like images from custom icons, or additional formatting via HTML. How do you propose storing that info in `data-title`?

**Comment by facelessuser** [comment]
Yeah, I don't think you can pass raw HTML in a `data-title`. You would not get the intended exact title in PDFs or print as far as I can tell.

**Comment by squidfunk** [comment]
That is indeed a problem. However, I just had the idea that we could use flex ordering and `display: contents` to achieve the same result. The idea is to move the elements one level up using `display: contents` on `.tabbed-labels` and `.tabbed-content`, set `flex-direction: column` on the top-level flex container and then use `order: $i` to put labels and containers in the right order:

<img width="1431" alt="Bildschirmfoto 2021-08-08 um 17 30 39" src="https://user-images.githubusercontent.com/932156/128637330-5c60549a-cc73-4158-96a0-33838b068ca6.png">

**Comment by facelessuser** [comment]
Cool. Look forward to checking it out.

**Comment by squidfunk** [comment]
Done!

<img width="1226" alt="Bildschirmfoto 2021-08-08 um 17 59 10" src="https://user-images.githubusercontent.com/932156/128638096-c58349c6-9423-4dfb-9607-1e9da9c6c478.png">

**Comment by squidfunk** [comment]
I've tested it in Chrome, Firefox and Safari - all good. I was a little concerned regarding `display: contents`, because Safari has [severe a11y bugs according to caniuse](https://caniuse.com/css-display-contents), but those don't matter in the print view. So, I'd say, looks pretty good!

**Comment by squidfunk** [comment]
Regarding naming: I find `content` and `subcontent` a little too unspecific. Here's my proposal:

``` html
<div class="tabbed-set" data-tabs="1:2">
  <input ...>
  <input ...>
  <div class="tabbed-labels">
    <label for="__tabbed_1_1">Foo</label>
    <label for="__tabbed_1_2">Bar</label>
  </div>
  <div class="tabbed-container">
    <div class="tabbed-content">
      ...
    </div>
    <div class="tabbed-content">
      ...
    </div>
  </div>
</div>
```

Alternatively, if you'd like to keep `.tabbed-content`, we could rename `.tabbed-subcontent` to `.tabbed-block`, and add a plural `s` to `.tabbed-content[s]`? It would be more consistent with `.tabbed-labels`.

``` html
<div class="tabbed-set" data-tabs="1:2">
  <input ...>
  <input ...>
  <div class="tabbed-labels">
    <label for="__tabbed_1_1">Foo</label>
    <label for="__tabbed_1_2">Bar</label>
  </div>
  <div class="tabbed-contents">
    <div class="tabbed-block">
      ...
    </div>
    <div class="tabbed-block">
      ...
    </div>
  </div>
</div>
```

I'm also open to other ideas, but I'm not a fan of `.tabbed-subcontent`.

**Comment by facelessuser** [comment]
I was mainly trying to reuse names of the legacy method to cut down on even more abstraction needed. I think renaming `.tabbed-subcontent` to `.tabbed-block` would allow me to keep everything else basically the same, so I prefer this.

**Comment by squidfunk** [comment]
That's cool! I think it's much better than `subcontent` and matches the nature of the element, as it's a block.

**Comment by squidfunk** [comment]
So, when you found the time (no hurry) to evaluate the solution I propose in the PR, I'd say we could follow the plan outlined in  https://github.com/facelessuser/pymdown-extensions/issues/1415#issuecomment-894751702. I'd update the [recommended Tabbed configuration](https://squidfunk.github.io/mkdocs-material/reference/content-tabs/?h=tabbed#configuration) in the docs to include `alternate_style: true`, so new users will pick it up. If we find no blockers in the coming months, we could at some point, switch implementations, and remove the old implementation. Other than that, if you favor to keep the old style, I would only remove it from Material for MkDocs and keep the `alternate_style` configuration flag.

**Comment by facelessuser** [comment]
Before I can make a release, I would also need to provide a simple CSS setup for non-Material people. The alternate style, while tailored as an alternative for Material, must be documented in such a way that anyone could make use of it. If not, I'd have to push it to the "material extra" plugin.

**Comment by squidfunk** [comment]
That makes sense! I thought we first test it on Material for MkDocs before it is publicly announced, but I can totally understand if you want to chime in non-Material users how to use it 😊

**Comment by facelessuser** [comment]
We could keep it undocumented at first. Assuming this turns into something we wish to formally support long-term, If you're willing to help put together a basic CSS example for the future, that is basically all I'd need. It just has to be something generally accessible.

**Comment by squidfunk** [comment]
The easiest way is to take the CSS output of Material and remove the `.md-typeset` selector at the beginning and any globally defined CSS variables. The remaining CSS should be self-contained then.

**Comment by facelessuser** [comment]
Perfect, that gives me something to play with. I'm assuming, if someone technically wanted unlimited tabs, some of this could be done with JavaScript, right? I'm not asking for such a solution, but I am assuming we could note that that is a possibility and just leave it up to the user if they were so motivated?

**Comment by squidfunk** [comment]
I've updated the example and reduced it further. 

> Perfect, that gives me something to play with. I'm assuming, if someone technically wanted unlimited tabs, some of this could be done with JavaScript, right? I'm not asking for such a solution, but I am assuming we could note that as a possibility and just leave it up to the user if they were so motivated?

I think so. You could just set the `order` property as an inline style on the respective elements.

**Comment by facelessuser** [comment]
Cool. Can't wait to play with this more. With the info I have, I think the basic documentation is possible now. I'll play around with all of this and let you know when things are ready.

**Comment by squidfunk** [comment]
Awesome! I really love how this turned out and how we solved problems on the way. I think it will be a great improvement and superior to the current solution. Thanks for your continued attention to this project!

I will also revisit improving the UI even further with possibly indicating the possibility to scroll via chevrons or something similar. I'll play around with it after we managed to push it out.

**Comment by facelessuser** [comment]
Renamed `tabbed-subcontent` to `tabbed-block`.

I did notice that sometimes when you have multiple tabs, and the tab to right is overflowing, it won't snap into view fully when selected.

**Comment by facelessuser** [comment]
> I did notice that sometimes when you have multiple tabs, and the tab to right is overflowing, it won't snap into view fully when selected.

This seems to not be a real big deal. It's not always reproducible, and it seems to occur with only a slight overflow of the active tab. It's just a quirk of the browser and nothing that could really be "fixed". Overall the snapping seems to work good.

**Comment by facelessuser** [comment]
Overall, I like it. The one thing I don't is how tabs can get hidden with overflow. I know with JS we could add some pagination buttons on overflow, but without them, it can be hard to tell sometimes that there are more tabs.

![tabs](https://user-images.githubusercontent.com/1055125/128806098-aaf54a39-22e0-461e-90ef-13b0baa2586e.gif)

**Comment by squidfunk** [comment]
Thanks for testing!

> I did notice that sometimes when you have multiple tabs, and the tab to right is overflowing, it won't snap into view fully when selected.

If you could manage to provide a reproducible case, I can look into it. It might be related to `scroll-snap-type:  x proximity`, which will not always snap to grid. Maybe we should remove scroll snapping entirely, I'm yet undecided.

> Overall, I like it. The one thing I don't is how tabs can get hidden with overflow. I know with JS we could add some pagination buttons on overflow, but without them, it can be hard to tell sometimes that there are more tabs.

That's basically the same problem as already exists for code blocks and tables, which also overflow on mobile, or anything else that is not wrappable in its entirey. We could stretch the `.tabbed-labels` container to the full width on mobile, as we did for code blocks and tables, which should better indicate potentially overflowing content. That's also what GitHub does:

<img width="422" alt="Bildschirmfoto 2021-08-10 um 09 53 04" src="https://user-images.githubusercontent.com/932156/128829794-646e637d-a86d-4e06-a7e0-8d89ffbc87cf.png">

Unfortunately, indicating overflow, there is not `:overflows` pseudo selector for CSS, so it has to be progressively enhanced with JavaScript, but I think this is something we could tackle after we got the first feedback.

**Comment by facelessuser** [comment]
> That's basically the same problem as already exists for code blocks and tables

Yep, but they do have visual indicators that things are overflowing in these two cases (as you mentioned). It makes all the difference when talking about intuitiveness.  Is it always the most attractive? Maybe not, but I can tell it is overflowing.

> Unfortunately, indicating overflow, there is not `:overflows` pseudo selector for CSS, so it has to be progressively enhanced with JavaScript, but I think this is something we could tackle after we got the first feedback.

Yep, and I am aware that JS is the only way. We'll probably never see an `:overflows` pseudo selector either as I can see people accidentally creating infinite loops with such a selector. Ah, we are overflowing, let's make it not overflow (removes overflow state which then removes overflow prevention style), loop ♾️.

> If you could manage to provide a reproducible case, I can look into it. It might be related to `scroll-snap-type: x proximity`, which will not always snap to grid. We could also set it to `x mandatory`, which however would not allow labels that exceed the screen width entirely, as it would always snap to the start. Maybe we should remove scroll snapping entirely, I'm yet undecided.

Honestly, I'm not really that concerned with this case. I initially noticed this, cycled through the different snap options, and settled that I think I like it better with than without.

> We could stretch the `.tabbed-labels` container to the full width on mobile, as we did for code blocks and tables, which should better indicate potentially overflowing content. That's also what GitHub does:

I may experiment with this. The visual indicator of overflow, even if the end result is not aesthetically the best is probably more important, but I need to see how bad it looks first 🙃 .

**Comment by facelessuser** [comment]
The only other thing I can think of is having the tab line expand past the tabs, so when you select a tab, the highlight underline won't reach the far left or far right, so there is a visual indicator that you aren't fully left or fully right. I'm not saying it's great, but you can see when you are not at the edge.

![tabs2](https://user-images.githubusercontent.com/1055125/128870428-7e359583-ce92-40ca-ac82-43c17f75b54c.gif)

**Comment by squidfunk** [comment]
> The only other thing I can think of is having the tab line expand past the tabs, so when you select a tab, the highlight underline won't reach the far left or far right, so there is a visual indicator that you aren't fully left or fully right.

That's a clever idea, but I think it looks a little off. I would go for removing the left and right margin on mobile (as GitHub) and add visual indicators with JS that tabs are overflowing and more tabs hide behind that. In the end, the user has to learn that tabs are scrollable. However, I would like to collect some feedback first. Sometimes somebody else comes up with a better idea.

Another idea - the background attachment hack:
https://lea.verou.me/2012/04/background-attachment-local/

**Comment by facelessuser** [comment]
Yeah, since I indent text and such to the start of the tab label, it actually doesn't look too bad:

<img width="312" alt="Screen Shot 2021-08-10 at 7 01 41 AM" src="https://user-images.githubusercontent.com/1055125/128871611-7fa5e09a-51bb-439c-ba82-5150807b5f71.png">

But yeah, it makes sense to get more ideas too.

**Comment by squidfunk** [comment]
Overall, I think the new implementation is so much better than what we currently have, because when tabs break it appears to be completely broken. There's nearly no reminiscence that those are tabs when they break onto multiple lines.

**Comment by facelessuser** [comment]
I agree that the behavior is better. I may use the extended line method in my personal documentation (disabling snapping to ensure that extended inline is always seen at first) as the default unless/until a better visual indicator comes along. I just don't want occasions to arise where a person misses (especial on mobile) a useful tab.

**Comment by squidfunk** [comment]
The latest commit on the feature branch now stretches the tab labels container just like code blocks when the viewport is below 480px:


https://user-images.githubusercontent.com/932156/128875342-06044296-519c-4cab-ae49-24512838d32a.mp4

Furthermore, `scroll-padding` ensures that tabs always align correctly.

**Comment by facelessuser** [comment]
As far as I can tell (after doing some research as well) what you've done is probably the best pure CSS tab implementation that really can be done. I have some other feature things I need to finish for the next release, so I'll probably look into those while I think about whether I turn on the new style by default in pymdown docs and, if so, what styling tweaks I settle on to make overflow tabs noticeable everywhere.

I think I've done everything in this pull needed to enable the feature. I'll document and settle on wether to turn it on and what style to use once I have everything else ready for the next release.

**Comment by squidfunk** [comment]
Perfect! As soon as you pushed out the new release, I'll adjust my PR (mainly to rename `.tabbed-subcontent` to `.tabbed-block` as discussed) and will also issue a new release. I'll add the new configuration flag as the canonical way to enable tabs to the docs. In the next major release, if no show-stoppers are encountered, I'll drop support for the legacy implementation.

**Comment by squidfunk** [comment]
Looks great! The HTML stayed the same? I'll look into it as soon as I find some time.

**Comment by facelessuser** [comment]
Yeah, HTML didn't change. Some way of knowing that there were more tabs was my main concern. I tweaked the scroll JS a little and probably some CSS. The minimal example is all demonstrated here in this codepen: https://codepen.io/facelessuser/pen/VwWdBQX.

I imagine if we wanted functional buttons that did something more than indicate "hey, there's more here" on mouseover/link click on mobile, then maybe something would have to change. I needed something decent enough to release with and figured you'd probably do something similar or propose something even better. It's marked as experimental, so if I need to change something later, that's fine.
